core/execution/future/mark_price.go

```diff
@@ -86,38 +94,101 @@ func NewCompositePriceCalculatorFromSnapshot(mp *num.Uint, mpc *snapshot.Composi
 		bookPriceAtTime[tp.Time] = num.MustUintFromString(tp.Price, 10)
 	}
 
-	return &CompositePriceCalculator{
+	calc := &CompositePriceCalculator{
 		config:           config,
 		trades:           trades,
 		sourceLastUpdate: mpc.PriceSourceLastUpdate,
 		priceSources:     priceSources,
 		bookPriceAtTime:  bookPriceAtTime,
 		price:            compositePrice,
+		timeService:      timeService,
 	}
+
+	if len(config.DataSources) > 0 {
+		oracles := make([]*products.CompositePriceOracle, 0, len(config.DataSources))
+		for i, s := range config.DataSources {
+			oracle, err := products.NewCompositePriceOracle(ctx, oe, s, config.SpecBindingForCompositePrice[i], calc.GetUpdateOraclePriceFunc(i))
+			if err != nil {
+				return nil
+			}
+			oracles = append(oracles, oracle)
+		}
+		calc.oracles = oracles
+	}
+	return calc
 }
 
-func NewCompositePriceCalculator(config *types.CompositePriceConfiguration) *CompositePriceCalculator {
+func NewCompositePriceCalculator(ctx context.Context, config *types.CompositePriceConfiguration, oe products.OracleEngine, timeService excommon.TimeService) *CompositePriceCalculator {
 	priceSourcesLen := len(config.SourceStalenessTolerance)
 	if priceSourcesLen == 0 {
 		priceSourcesLen = 1
 	}
 
-	return &CompositePriceCalculator{
+	mpc := &CompositePriceCalculator{
 		config:           config,
 		priceSources:     make([]*num.Uint, priceSourcesLen),
 		sourceLastUpdate: make([]int64, priceSourcesLen),
 		bookPriceAtTime:  map[int64]*num.Uint{},
+		timeService:      timeService,
+	}
+	if len(config.DataSources) > 0 {
+		oracles := make([]*products.CompositePriceOracle, 0, len(config.DataSources))
+		for i, s := range config.DataSources {
+			oracle, err := products.NewCompositePriceOracle(ctx, oe, s, config.SpecBindingForCompositePrice[i], mpc.GetUpdateOraclePriceFunc(i))
+			if err != nil {
+				return nil
+			}
+			oracles = append(oracles, oracle)
+		}
+		mpc.oracles = oracles
 	}
+	return mpc
 }
 
-func (mpc *CompositePriceCalculator) UpdateConfig(config *types.CompositePriceConfiguration) {
+func (mpc *CompositePriceCalculator) UpdateConfig(ctx context.Context, oe excommon.OracleEngine, config *types.CompositePriceConfiguration) error {
+	// special case for only resetting the oracles
+	if config == nil && mpc.oracles != nil {
+		for _, cpo := range mpc.oracles {
+			cpo.UnsubAll(ctx)
+		}
+		return nil
+	}
+
 	priceSourcesLen := len(config.SourceStalenessTolerance)
 	if priceSourcesLen == 0 {
 		priceSourcesLen = 1
 	}
 	mpc.config = config
 	mpc.priceSources = make([]*num.Uint, priceSourcesLen)
 	mpc.sourceLastUpdate = make([]int64, priceSourcesLen)
+	for _, cpo := range mpc.oracles {
+		cpo.UnsubAll(ctx)
+	}
```

Comment:
Could / should this be changes by that? As it is now it looks like both for loop are the same, but being duplicated, we could miss updating one or the other in the future if needed. At the end of the day I don't mind much.
```suggestion
	for _, cpo := range mpc.oracles {
		cpo.UnsubAll(ctx)
	}
	if config == nil && mpc.oracles != nil {
		return nil
	}

	priceSourcesLen := len(config.SourceStalenessTolerance)
	if priceSourcesLen == 0 {
		priceSourcesLen = 1
	}
	mpc.config = config
	mpc.priceSources = make([]*num.Uint, priceSourcesLen)
	mpc.sourceLastUpdate = make([]int64, priceSourcesLen)

```

Reply:
something like that, yeah, can do.